system/mp/mpcomm.cpp

```diff
@@ -441,97 +444,321 @@ class CMPServer;
-            // But this is all probably OTT. The main thing is to avoid a slow client blocking the accept loop.
-            while (true)
+            if (socketHandler.queryReadSoFar()) // read something
             {
-                sem.wait();
-
-                Owned<ISocket> sock;
+                VStringBuffer errMsg("MP Connect Thread: invalid number of connection bytes serialized from: %s", socketHandler.queryPeerHostText());
+                FLLOG(MCoperatorWarning, "%s", errMsg.str());
+            }
+            int exceptionCode = exception->errorCode();
+            switch (exceptionCode)
+            {
+                case JSOCKERR_timeout_expired:
+                    trackPortProbe(get_cycles_now() - socketHandler.queryCreateTime(), socketHandler.queryPeerEndpointText(), CloseType_timeout);
+                    break;
+                case JSOCKERR_graceful_close:
+                    trackPortProbe(get_cycles_now() - socketHandler.queryCreateTime(), socketHandler.queryPeerEndpointText(), CloseType_graceful);
+                    break;
+                default:
+                    trackPortProbe(get_cycles_now() - socketHandler.queryCreateTime(), socketHandler.queryPeerEndpointText(), CloseType_error);
+                    break;
+            }
 
-                {
-                    CriticalBlock b(crit);
-                    if (stopped)
-                        break;
-                    if (slowClientsSocks.empty()) // guard, but should never happen
-                    {
-                        WARNLOG("slowClientsSocks list empty");
-                        continue;
-                    }
-                    sock.set(slowClientsSocks.back());
-                    slowClientsSocks.pop_back();
-                }
-                owner.handleAcceptedSocket(sock.getClear(), CONFIRM_TIMEOUT, true);
+            Linked<CSocketHandler> handler = &socketHandler;
+            {
+                CriticalBlock b(handlersCS);
+                selectHandler->remove(socketHandler.querySocket());
+                handlers.remove(&socketHandler);
+            }
+            handler->querySocket()->close();
+        }
+        void process(CSocketHandler &socketHandler)
+        {
+            Linked<CSocketHandler> handler = &socketHandler;
+            {
+                CriticalBlock b(handlersCS);
+                selectHandler->remove(socketHandler.querySocket());
+                handlers.remove(&socketHandler);
             }
+
+            if (owner.threadPool)
+                owner.threadPool->start(handler.getClear());
+            else
+                owner.handleAcceptedSocket(handler.getClear());
         }
-    } slowClientProcessor;
+    };
+    std::atomic<bool> running;
+    bool listen;
+    ISocket *listensock;
+    CMPServer *parent;
+    int mpSoMaxConn;
+    unsigned acceptThreadPoolSize = 0;
+    unsigned maxListenHandlerSockets = 60000; // what is a sensible default limit?
```

Comment:
Just a comment, if you expect a larger number of fds then when creating the selectHandler you can specify an argument to it for how many fds it handles before creating a new thread for more:
createSocketSelectHandler(traceName, fdsPerThread)

Reply:
I think we should leave that for a separate discussion/change. It (MP) has always used the default (UINT_MAX) and hence 1 CSocketEpollThread. Actually looks like everywhere uses the default. May be worth revisiting to see if it helps when a lot, but separate issue.

This limit (maxListenHandlerSockets), is limiting the number of clients that are actively connecting at once - once they're actually connected they are processed by handleAcceptedSocket and added to the standard MP select handler.
Not sure what a sensible # is here, perhaps much lower, but it's here to catch a crazy burst of connections coming in (like a DoS).